### DISCUSSION TITLE
在搜索功能上线前考虑对挂载的其他 AList 站点的保护

### DISCUSSION THREAD

**Discussion by BoYanZh** [AlistGo/alist#2529] [General]
在搜索功能上线前需要考虑如何保护其他 AList 站点。由于建立索引需要遍历目录，而如xiaoya，qingyuan，chirmyram这样的站点被其他用户以webdav或者alist的driver挂载数可能很多，遍历目录时可能直接把站点打崩。以及可能出现的循环挂载而用户建索引时没有忽略掉循环的路径导致大量无效请求，也可能使一批站点被封甚至 AList 被拉入恶意程序黑名单。故需要考虑相关保护机制。

根据群里的讨论大概有如下几个设想：

1. 建索引时直接跳过 driver 为 AList 的路径。（较容易实现）
2. 让 AList 站长提供选项能否被其他 AList 挂载，AList 在请求其他 AList 站点时带上特殊 header 便于判断。更理想的情况是正常浏览和建立索引时使用不同 header，给 AList 站长更细粒度的控制。（在目前的架构下中需要比较大的改动）
3. 考虑提供一个列出 driver 为 AList 的接口，在建立索引前先去请求 driver 为 AList 的 storage 的该接口，然后自动加到忽略的路径中。
4. 设计侦测循环挂载的算法，检测到循环直接停止，但较难实现而且这种不太容易出现的情况可能较大地影响效率。同时也没法给别人的 AList 站点加锁，导致在建索引过程中有路径变化可能会识别失败。

2和3两个想法会受制于某些的站点的 AList 程序没有更新而不会有这个功能，导致没法达到预料中的结果，需要注意。

**Comment by BoYanZh** [comment]
2和3也可以反向实现，如果想挂载的AList没有明确说明可以挂载则不去挂载

**Comment by BoYanZh** [comment]
个人现在觉得最可行的办法是alist默认可以被挂载，不可以被搜索时建索引。新版本提供专门的接口返回这两个值，如果不是新版本就使用默认值，并提供前端给管理员设置。在`walkFS`的时候遇到`stoarge`为`AList V2`或`AList V3`的时候就先去请求这个接口再决定要不要去爬取。

**Comment by xhofe** [comment]
可以先从最简单的实现开始

**Reply by iemobiu** [reply]
赞同，支持。目前v3.5.1页面右上角的搜索还不能用吧？

**Reply by nxjonson** [reply]
同样3.5.1版本，右上角没有“搜索”，请问需要如何设置？

**Reply by Byz-liu** [reply]
现在别人不需要令牌就可以挂载自己的alist，这样自己辛辛苦苦整理保存的资源别人轻轻松松就能挂载而且很多人还不注明来源，现在只能对自己的文件夹就行加密然后readme写明密码，这样来访问的人也很麻烦每次都要输密码才能看见。

能不能实现只有令牌的人才能挂载自己的alist，或者添加一个授权域名，为每个域名分配权限比如能不能挂载隐藏文件或者加密文件能不能索引等。如果能实现或者有其他的解决方法万分感谢！

**Reply by zhx47** [reply]
或许也可以加个选项，是否将整个索引也一起共享出去

**Comment by iemobiu** [comment]
<img width="437" alt="截屏2022-12-03 00 22 08" src="https://user-images.githubusercontent.com/115678292/205338388-85d88f29-c234-47a3-b150-becd4dbc6c21.png">
刚简单的测试了下搜索功能，还挺好用。但配置那里，“Current indexes”为何一直转圈呢
database和bleve的区别是什么呢？
目前有关于搜索功能的比较详细的配置说明吗？

**Reply by iemobiu** [reply]
点了别的链接，然后又点了搜索这页面，不转圈了，出现了对号。貌似自己不会停下来，需要手动刷新页面。是不是这个功能还在测试中。
